lib/parallel_test_base.pm

```diff
@@ -0,0 +1,1946 @@
+Check source and destination hosts operating system version. Guest migration can
+not be done from the newer to the older. Main argument is host role (src or dst).
+=cut
+
+sub check_host_os {
+    my ($self, %args) = @_;
+    $args{role} //= 'src';
+
+    record_info("Check host os");
+    my $_ret = 0;
+    my $_localip = get_var('LOCAL_IPADDR');
+    my $_peerip = get_var('PEER_IPADDR');
+    if (is_sle) {
+        my ($_localosver, $_localossp,) = get_os_release;
+        my ($_peerosver, $_peerossp,) = get_os_release("ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_peerip");
+        print "LOCAL OS $_localosver-sp$_localossp PEER OS $_peerosver-sp$_peerossp\n";
+        save_screenshot;
+        unless (($_peerosver > $_localosver) or ($_peerosver == $_localosver and $_peerossp >= $_localossp)) {
+            $_ret = 1;
+            if ($args{role} eq 'src') {
+                croak("Destination os $_peerosver-sp$_peerossp falls behind source os $_localosver-sp$_localossp");
+            }
+            elsif ($args{role} eq 'dst') {
+                record_info("Source os $_peerosver-sp$_peerossp  falls behind destination os $_localosver-sp$_localossp");
+            }
+        }
+    }
+    return $_ret;
+}
+
+=head2 check_host_virtualization
+
+Check virtualization modules and services are ready, otherwise test run can not
+proceed.
+=cut
+
+sub check_host_virtualization {
+    my $self = shift;
+
+    record_info("Check host virtualization");
+    if (is_kvm_host) {
+        assert_script_run("lsmod | grep kvm");
+    }
+    elsif (is_xen_host) {
+        assert_script_run("lsmod | grep xen");
+    }
+
+    if (script_run("systemctl is-active libvirtd") != 0) {
+        systemctl("stop libvirtd", ignore_failure => 1);
+        systemctl("start libvirtd");
+        systemctl("is-active libvirtd");
+    }
+    save_screenshot;
+}
+
+=head2 check_host_network
+
+Check source and destination hosts are in the same subnet, otherwise test run
+can not proceed.
+=cut
+
+sub check_host_network {
+    my $self = shift;
+
+    record_info("Check host network");
+    my $_localgw = script_output("ip route show default | awk \'{print \$3}\'");
+    my $_localip = get_var('LOCAL_IPADDR');
+    my $_peerip = get_var('PEER_IPADDR');
+    my $_peergw = script_output("ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_peerip ip route show default | awk \'{print \$3}\'");
+    save_screenshot;
+    croak("$_localip and $_peerip hosts are not in the same subnet") unless ($_localgw eq $_peergw);
+}
+
+=head2 check_host_package
+
+Install necessary packages to facilitate test run down the road. Main argument
+is packages to be installed.
+=cut
+
+sub check_host_package {
+    my ($self, %args) = @_;
+    $args{package} //= '';
+
+    record_info("Check host package");
+    zypper_call("--gpg-auto-import-keys ref");
+    is_kvm_host ? zypper_call("in -t pattern kvm_tools") : zypper_call("in -t pattern xen_tools");
```

Comment:
This won't work for alp... It is ok to be extended in new PR later, but better keep in mind the difference about "host" between alp and sle, and add some if ... else or TODO comment ahead of time, to save some redesign/extend effort.

Reply:
IF...ELSE can be added when needed.